### ISSUE TITLE
RNA-seq builds are incorrectly using only 1 CPU

### ISSUE THREAD

**Issue by malachig** [genome/gms#48]
For some reason, RNA-seq jobs run on the standalone GMS are using only 1 CPU during the alignment step (instead of 12).  The lsf requested resources for the job look like this:

Requested Resources 

```
<select[ncpus>=12 && mem>=28672] span[hosts=1] rusage[mem=28672]>;
```

However, the alignment log file states that 'CPU count is 1', and is running:
/usr/lib/bowtie2.0.0-beta7/bin:$PATH /usr/bin/tophat2.0.4 --num-threads 1 ...

Top confirms that only a single CPU is being used.  This will take an extremely long time to complete.

It possible that this issue is caused by a customization to lsf resource requests in open lava to remove the ncpus option.  We did that because it was preventing jobs from starting at all, but perhaps the RNA-seq job needs this to correctly determine the number of CPUs to use.  How does the RNA-seq job set this thread number?

**Comment by malachig** [comment]
It looks like the number of CPUs to use in an RNA-seq build is determined by a call to 

```
$self->_available_cpu_count
```

in InstrumentData/AlignmentResult/PerLaneTophat.pm.  This method actually lives in SoftwareResult.pm.

This method seems to set the number of CPUs to 1 as follows: 

```
if (!exists $ENV{LSB_MCPU_HOSTS}){
    return 1;
}
```

Not sure yet where this environment variable is defined in TGI.  Not in Site/TGI.pm.  And not globally defined on my linux box.  Perhaps this is actually configured on the LSF blades themselves?  Or perhaps this environment variable is defined on the blade according to the resource request made to LSF?

In any case it seems that it would not be set in the standalone GMS.  

We could define this in /etc/genome.conf perhaps?    Perhaps by defining a new environment variable $OPENLAVA_MCPU_HOSTS and adding logic to 'available_cpu_count()' in SoftwareResult.pm to handle the openlava/standalone gms situation...

We should probably confer with Matt, Jason, and perhaps some Apipe folks to better understand what is happening here before pushing a fix for this one...

In any case, I think we should make it more clear what is happening by warning the user that $LSB_MCPU_HOSTS is not defined and that only 1 cpu will be used before silently returning 1.

**Comment by gatoravi** [comment]
So 'bwa' seems to be using similar logic, 
In InstrumentData/AlignmentResult/Bwa.pm there are 2 lines which look like, 

 `my $cpu_count = $self->_available_cpu_count;
 $cpu_count = $cpu_count * 20;`

This sets the number of threads to 20 for the 'bwa' aln step and this can be seen in the logs of the builds.

 But a similar line( $cpu_count = $cpu_count \* 20;) does not exist for TopHat, would it be as simple as adding this line ?

**Comment by malachig** [comment]
Hmmm,

That is sub-optimal.  We need a better way to control this than just multiplying by 20.  Apparently this change is specific to the genome 'pub' branch.  This explains the unexpected high cpu usage I have observed in the past during bwa alignments.  I expect Scott previously encountered a similar problem where bwa was running with only 1 thread (as we are seeing in rna-seq).  This may also be related to issue: #10 

We need to figure out why we are not able to parse the LSF resources for an LSF job in the standalone GMS.  

If we can get the number of CPUs actually requested of LSF, we should then use that.  Right now there may be many places in the pipeline where availablecpu_count() is called and is incorrectly returning 1 and therefore a process that is expected to be run in parallel is running only with 1 thread.

If deemed necessary we should improve the warnings in the method to make it more clear that we failed to obtain the cpu_count and are running with only 1 thread.

**Comment by malachig** [comment]
It does seem that $ENV{LSB_MCPU_HOSTS} is defined in lsf jobs run by openlava on the clia1 box

```
% bsub -M 20000000 -n 4 -R 'select[ncpus>=8 && mem>20000] rusage[mem=20000] span[hosts=1]' 'env > /tmp/envs.txt'
% cat /tmp/envs.txt | grep -i LSB
```

Gives this:
LSB_MCPU_HOSTS=clia1 4

It appears that this variable gets set according to what is passed into lsf with the '-n' option.  Why is only 1 cpu being found and used in tophat?

We need better logging of the actual lsf commands being submitted and the value of $LSB_MCPU_HOSTS obtained within the process...

**Comment by gatoravi** [comment]
Since '-n' is disabled by the commit mentioned above LSB_MCPU_HOSTS is set to 1.

Just to summarize, LSB_MCPU_HOSTS is set by the bsub -n parameter. InstrumentData/AlignmentResult/PerLaneTophat.pm uses this variable to set the number of threads that TopHat uses.

Also 'ncpus' is used to request the number of cores/node.

**Comment by malachig** [comment]
reposting comment from #10 for reference,

OK, i think the commit that Scott( or Obi?) made is this one,
genome/tgi-workflow@dbd055c

Looks like the ncpus option is still used but -n is disabled for OPENLAVA.

```
if (defined $job->resource->min_proc) {
        unless ($OPENLAVA) {
        $cmd .= sprintf("-n %s ", $job->resource->min_proc);
}
```

**Comment by malachig** [comment]
I think we should revert this change.  And start allowing the '-n' option to be used.  It seems there are many places where this option is being parsed via $LSB_MCPU_HOSTS to determined the number of threads to use.

We should also revert the manual change in Bwa.pm to use 20 cores in the standalone GMS.

If we have problems with jobs not launching on systems with few processors (e.g. a virtual machine running on a desktop), perhaps we can introduce a new environment variant specific to the GMS called 'GENOME_MAX_CPUS' that could be set in /etc/genome.conf.

If defined we could have code in lib/Workflow/Dispatcher/Lsf.pm that overrides the '-n' option to use this number of cpus.

To be totally clear.  
1.) Remove the described snippet above from:
Workflow/Dispatcher/Lsf.pm  

2.) Replace it with logic to check for $ENV{GENOME_MAX_CPUS} and $job->resource->min_proc and use whichever is lower for the '-n ' option.  Only do this if $ENV{GENOME_MAX_CPUS} is actually defined.

3.) Remove the hack from InstrumentData/AlignmentResult/Bwa.pm that overrides the command to use 20 cpus.

4.) In the GMS installation repo, edit: setup/etc/genome.conf.  Add GENOME_MAX_CPUS with a default value of 2 but have it commented out by default.

**Comment by malachig** [comment]
According to Matt we may have an issue with our openlava configuration.  

If PARALLEL_SCHED_BY_SLOT=Y in lsb.params, this option specifies the number of slots required to run the job, not the number of processors.  What I do not know is… how lsf (openlava) perceives the difference between a "processor" and a "processor core"

We can try to add this configuration to:
/opt/openlava-2.0/etc/lsb.params

Then restart lsf by:
sudo /etc/init.d/openlava restart

#Test this in the standalone GMS and if it works figure out a way to get this configuration added automatically during the make process

UPDATE: 
Unfortunately this did not work.  PARALLEL_SCHED_BY_SLOT is not considered a valid configuration value by openlava.

This now seeming more and more like an openlava vs. lsf problem.  The solution in my previous will not work very well.  If we allow the '-n' option again, various places in the Genome code will use this (via $LSB_MCPU_HOSTS) to set the number of threads for various jobs.  This would be fine except in openlava, '-n' is interpreted as the number of physical processors (not cores or threads).  So the default number used for '-n' in tophat alignments will be 12 and this will be unstartable according to openlava on almost any machine.

Perhaps we can maintain the openlava specific logic that drops use of the '-n' option, and then find some way to parse the 'ncpus' resource request supplied to openlava via the '-R' option.

**Comment by malachig** [comment]
Workflow has logic to handle OpenLava in Workflow/Dispatcher/Lsf.pm, around line 77:

``` perl
    77     if ($OPENLAVA and @rusages > 1) {
    78         my @new = ($rusages[0]);
    79         shift @rusages;
    80         warn "Discarding part of rusaage for OpenLava.  Keep: @new.  Discard: @rusages.\n";
    81         @rusages = @new;
    82     }
    83 
    84     my $rusage = join(", ", @rusages);
```

This is done because OpenLava dosn't digest the comma syntax.  There may be another way to give OpenLava a complex rusage, but the current implementation ignores all but the first part of the resource request.

**Comment by malachig** [comment]
Our current attempt to solve the problem of getting the number of CPUs from openlava.
- First, do not use the -n option in openlava.  This is specified in Workflow/Dispatcher/Lsf.pm.  There are a few openlava specific tweaks there.  One removes the '-n' option when crafting job submissions.  Another removes attempts to use a 'gtmp' resource.  And another truncates multiple key/value pairs in an rusage statement to just first key/value pair.
- Second, we have added some conditional logic to Genome/SoftwareResult.pm to attempt to parse the number of cpus from the -R 'select[]' statement instead of the environment variable that is defined by specifying '-n'.  This is done by obtaining the lsf $job_id, running bjob -l $job_id and parsing the result.

**Comment by malachig** [comment]
Assuming the strategy above works, what will remain is the unfortunate situation of dropping all but the first resource request made in the rusage[] block (not to be confused with the select[] block).

It appears that openlava does not support rusage syntax with comma delimited lists of key/value pairs. This is why we dropped to using only the first key/value pair.  However, according to documentation here:
http://rpm.pbone.net/index.php3/stat/45/idpl/23854526/numer/1/nazwa/lsfintro#lbAD

"The syntax of the resource usage string is"res[=value]:res[=value]: ... :res[=value]" where 'res' is one of the resources whose value is returned by lsload. For example, "r1m=0.5:mem=20:swp=40" indicates that the task is expected to increase the 1-minute run queue length by 0.5, consume 20 Mbytes of memory and 40 Mbytes of swap space. If no value is specified, the task is assumed to be intensive in using that resource. In this case no more than one task will be assigned to a host regardless of how many CPUs it has.

The default resource usage for a task is assumed to be "r15s=2.0:r1m=2.0:r15m=2.0" which indicates a CPU intensive task which consumes few other resources."

In our test environment, valid values for rusage appear to be:
r15s   r1m  r15m   ut    pg  ls    it   tmp   swp   mem

In other words, it is possible that simply swapping the delimiter ", " with ":" will allows rsuage requests for lsf to work in openlava.  This change would occur in Workflow/Dispatcher/Lsf.pm and would only happen is $OPENLAVA was true.

**Comment by malachig** [comment]
@gatoravi pushed a fix for the comma vs. colon delimiter issue with resource requests to both master and gms-pub branches of tgi-workflow here:
https://github.com/genome/tgi-workflow/commit/e0bc4b4676d708bc84ffc943063df57c15c8b18c
https://github.com/genome/tgi-workflow/commit/58395d7a98e951090d42dc5b97635c0266a7f251

**Comment by gatoravi** [comment]
This fix seems to be working, bowtie now takes up 12 cores and bwa takes up 4 threads.

**Comment by malachig** [comment]
We may need further tweaks to how resource requests are made when using openlava instead of lsf and/or when we are running in a standalone environment with limited nodes and memory...   However, this specific issue seems to be resolved so I am closing it.
